src/api.rs

```diff
@@ -105,495 +104,249 @@ impl CoverCrypt {
+        let (symmetric_key, encapsulation) = cc.encaps(policy, public_key, attributes)?;
+
+        // encrypt the metadata using the DEM
+        let metadata = metadata
+            .map(|metadata| -> Result<Vec<u8>, Error> {
+                metadata.try_to_bytes().map_err(|_| Error::ConversionFailed)
+            })
+            .unwrap_or_else(|| Ok(vec![]))?;
+        let metadata = cc.encrypt(&symmetric_key, &metadata, None)?;
+
+        Ok((
+            symmetric_key,
+            Self {
+                encapsulation,
+                metadata,
+            },
         ))
-    })?;
-
-    // combine these values with all attribute values from the next axes
-    let other_values = combine_attribute_values(current_axis + 1, axes, map)?;
-    if other_values.is_empty() {
-        Ok(axis_values.iter().map(|v| vec![*v]).collect())
-    } else {
-        let mut combinations = Vec::with_capacity(axis_values.len() * other_values.len());
-        for av in axis_values {
-            for ov in &other_values {
-                let mut combined = Vec::with_capacity(1 + ov.len());
-                combined.push(*av);
-                combined.extend_from_slice(ov);
-                combinations.push(combined);
-            }
-        }
-        Ok(combinations)
     }
-}
 
-/// Convert an access policy used to decrypt ciphertexts into the corresponding
-/// list of CoverCrypt current partitions that can be decrypted by that access policy
-fn access_policy_to_current_partitions(
-    access_policy: &AccessPolicy,
-    policy: &Policy,
-) -> Result<HashSet<Partition>, Error> {
-    let attr_combinations = to_attribute_combinations(access_policy, policy)?;
-    let mut set = HashSet::with_capacity(attr_combinations.len());
-    for attr_combination in &attr_combinations {
-        for partition in to_partitions(attr_combination, policy)? {
-            let is_unique = set.insert(partition);
-            if !is_unique {
-                return Err(Error::ExistingCombination(format!("{attr_combination:?}")));
+    /// Decrypt the encrypted header with the given user secret key.
+    ///
+    /// - `cc`  : CoverCrypt scheme
+    /// - `usk` : CoverCrypt user secret key
+    pub fn decrypt(
+        &self,
+        cc: &CC,
+        usk: &CC::UserPrivateKey,
+    ) -> Result<ClearTextHeader<SYM_KEY_LENGTH, DEM>, Error> {
+        // decapsulate the symmetric key
+        let secret_keys = cc.decaps(usk, &self.encapsulation)?;
+
+        // Try to decrypt the metadata: CoverCrypt decapsulates of all the keys
+        // contained in an encapsulation with the user secret key. If the user
+        // is allowed to decapsulate, at least one of the keys returned by the
+        // `decaps` will be correct. The DEM scheme will fail for all other
+        // keys since it is authenticated.
+        for key in secret_keys {
+            if let Ok(plaintext) = cc.decrypt(&key, &self.metadata, None) {
+                return Ok(ClearTextHeader {
```

Comment:
just to understand, is that `cc.decrypt` very fast or could it be slow? I'm looking for improving performances here, as you have to decrypt (potentially) multiple times with different keys. How many keys are in `secret_keys`?

Reply:
Yes, this is a problem with te current implementation of CoverCrypt, we have to decrypt all the symmetric key encapsulations because our asymmetric scheme cannot detect the success of a decapsulation.

We could think about two solutions:
- EAKEM (cf [my other MR](https://github.com/Cosmian/cover_crypt/pull/39))
- the CCA property of the post-quantum KEM scheme KYBER (basically it decrypts the encrypted key, encrypts it again and verifies the ciphertext produced this way matches the one given as input; if it does not, the decapsulation failed), cf Algorithm 9 [in this paper](https://pq-crystals.org/kyber/data/kyber-specification-round3-20210804.pdf)

Reply:
There will be in `secret_keys` `n x m` keys, where `n` is the number of sets corresponding to the user secret key, and `m` the number of sets for which an encapsulation has been added in the CoverCrypt ciphertext.

Reply:
So in average, we will perform `n * m` asymmetric decryptions and `1/2 * n * m` symmetric decryption